### PR TITLE
Lab

### PR DESCRIPTION
**Pull request by andriikovall**

### CODE REVIEW

File: src/lib/euroDiffusion/isCountryCompleted.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { Country } from '../../types';
+import { EU } from '../prepareInput';
+
+export const isCountryCompleted = (country: Country, eu: EU): boolean => {
```
Comment (dpalii) [inline src/lib/euroDiffusion/isCountryCompleted.ts:4]: what about passing only country here and remove eu as an argument. I think it's possible. What do you think?

File: src/lib/euroDiffusion/sendCoinsToNeighbors.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { City } from '../../types';
+import { safeGetNumber } from '../../utils';
+
+export const sendCoinsToNeighbors = (city: City, neighbors: City[]) => {
+  for (const motif of Object.keys(city.coins.count)) {
+    const coinsRepresentativePortion = Math.floor(
```
Comment (dpalii) [inline src/lib/euroDiffusion/sendCoinsToNeighbors.ts:6]: let's move to to separate func

File: src/lib/prepareInput/prepareInput.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        coins: {
+          count: {},
+        },
+        coinsToReceive: {
+          count: {},
+        },
+        country: '',
+      }),
+    );
+    const countriesWithCities = countries.map(country => {
```
Comment (dpalii) [inline src/lib/prepareInput/prepareInput.ts:32]: let's move all this into one loop

File: src/lib/prepareInput/prepareInput.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    const matrix = createMatrix<City>(
+      MAX_X_COORDINATE + 1,
+      MAX_Y_COORDINATE + 1,
+      (x, y) => ({
+        x,
+        y,
+        coins: {
+          count: {},
+        },
+        coinsToReceive: {
```
Comment (dpalii) [inline src/lib/prepareInput/prepareInput.ts:26]: I think it's not th best way to handle coins, you will have to check the existance of key in the object and may have issues with it. What about simply having a `0` count for each of the motif by default?

File: src/lib/readInput/parseInput.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        ) ||
+        ylParsed > yhParsed
+      ) {
+        throw new Error(
+          `Invalid y coordinate: ${ylParsed} or ${yhParsed}. Min is ${MIN_Y_COORDINATE} and max is ${MAX_Y_COORDINATE}`,
+        );
+      }
+      countriesWithNoCities.push({
+        name,
+        xl: Number(xl),
```
Comment (dpalii) [inline src/lib/readInput/parseInput.ts:62]: `Number(xl)` can be replaced with `xlParsed` and so on. A small mistake, not a big deal

**Review by dpalii — COMMENTED** [review]
added several comments. Everything else looks pretty nice

**Comment by andriikovall** [comment]
updated all comments

**Pull request merged**